### PR TITLE
Fix  binary array to hex

### LINKED ISSUE #51: _binary_array_to_hex gives wrong value
Correct me if I am wrong, but `_binary_array_to_hex` gives wrong values.

```python
Python 2.7.13 (default, Feb 16 2017, 19:11:00) 
[GCC 5.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from PIL import Image
>>> import imagehash
>>> img = Image.open('sample_image.png')
>>> ph = imagehash.phash(img)
>>> bool_array = ph.hash.flatten()
>>> bool_array
array([ True, False, False,  True,  True,  True,  True, False, False,
       False,  True,  True,  True,  True, False, False,  True,  True,
       False, False, False, False, False,  True,  True,  True,  True,
        True, False, False, False, False,  True,  True,  True,  True,
       False, False, False, False,  True,  True, False, False, False,
       False,  True,  True,  True,  True,  True, False, False,  True,
        True,  True,  True,  True, False, False, False, False, False, False], dtype=bool)
>>> bit_array = 1*bool_array
>>> bit_array
array([1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0])
>>> bit_string = ''.join(str(b) for b in bit_array)
>>> bit_string
'1001111000111100110000011111000011110000110000111110011111000000'
>>> int(bit_string, 2)
11402201597170935744L
>>> hex(int(bit_string, 2))
'0x9e3cc1f0f0c3e7c0L' # This is the expected hex value for the given bit array (phash)
>>> str(ph)
'793c830f0fc3e703' # and here is the actual value.
```

If I made no mistake, let me know and I will submit a PR.

### PR DESCRIPTION
**Pull request by djunzu**
Fix _binary_array_to_hex as discussed in #51

Fixing _binary_array_to_hex demands hex_to_hash to be changed as well. That makes hex_to_hash incompatible with hex values generate with previous versions of ImageHash. For that reason, the old implementation is available as old_hex_to_hash in order to allow users convert old hex values back to hashes.

Because of that, I think that this PR should be followed by a major release and the README should point out how to upgrade.

Regarding the new version of hex_to_hash:
- It assumes all hashes are bidimensional arrays with dimensions hash_size * hash_size. This is the current behavior of all hashes.
- The implementation does not work for hash_size < 2. (But hash_size = 1 does not make sense anyway.

This PR also fix some tests.
This PR closes #51.

### CODE REVIEW

**Comment by JohannesBuchner** [comment]
That looks very good already, thank you!
What does the .python-version file do?

Could you also include the problematic cases you pointed out in #51 with Lenna.png (hash_size=6, hash_size=2) as new unit tests? You can use the imagehash.png in imagehash/tests/data/, which is not copyrighted.

Could you also include one or two unit tests for the old implementation, just to make sure we don't break it in the future (coverage decreased, I guess because of this).

I had been flirting earlier on with the idea of using hypothesis for testing (e.g. generating new images of various sizes, making hash of some size, generating string, make a hash out of this again). Not necessary to do this here, but might be an idea.

Do you think it would make sense to include bitarray (mentioned in #51) as a dependency?

**Comment by JohannesBuchner** [comment]
That looks very good already, thank you!
What does the .python-version file do?

Could you also include the problematic cases you pointed out in #51 with Lenna.png (hash_size=6, hash_size=2) as new unit tests? You can use the imagehash.png in imagehash/tests/data/, which is not copyrighted.

Could you also include one or two unit tests for the old implementation, just to make sure we don't break it in the future (coverage decreased, I guess because of this).

I had been flirting earlier on with the idea of using hypothesis for testing (e.g. generating new images of various sizes, making hash of some size, generating string, make a hash out of this again). Not necessary to do this here, but might be an idea.

Do you think it would make sense to include bitarray (mentioned in #51) as a dependency? Or binascii.hexlify/unhexlify? Or binhex.binhex/hexbin?

File: imagehash/tests/test_hex_conversions.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import imagehash
+
+
+# Each row is a test case where the first value is a bit sequence and
+# the second value is the expected hexadecimal representation for it.
+binary_to_hexadecimal_values = [
+	['1', '1'],
+	['11', '3'],
+	['111', '7'],
+	['1111', 'f'],
```
Comment (djunzu) [inline imagehash/tests/test_hex_conversions.py]: @JohannesBuchner , this test case covers hash_size=2. Previously there was no output. This test case assert there is a output and it is correct.

File: imagehash/tests/__init__.py [hunk 2]
```diff
@@ -41,24 +41,27 @@ def check_hash_algorithm(self, func, image):
...
-                '- stringified hash {}'.format(distance, image_hash,
-                                               other_hash))
-        self.assertEqual(distance, 0, emsg)
+    def check_hash_stored(self, func, image, sizes=range(2,21)):
+        for hash_size in sizes:
+            image_hash = func(image, hash_size)
+            other_hash = imagehash.hex_to_hash(str(image_hash))
+            emsg = 'stringified hash {} != original hash {}'.format(other_hash,
+                                                                    image_hash)
+            self.assertEqual(image_hash, other_hash, emsg)
```
Comment (djunzu) [inline imagehash/tests/__init__.py]: @JohannesBuchner , this test assert one can generate a hex from a hash and later generate the hash from the hex, for all values of hash_size >= 2 and < 21. It does not assert the hex value is correct, but it implicit test the problem with hash_size=6 described in #51.

**Comment by djunzu** [comment]
>  What does the .python-version file do?

I use [pyenv](https://github.com/pyenv/pyenv) to run different Python versions and it uses this file. It should not be committed, my fault. I will add it to .gitignore

> Could you also include the problematic cases you pointed out in #51 with Lenna.png (hash_size=6, hash_size=2) as new unit tests? 

It is already there. (See my comments in "files changed".)

> Could you also include one or two unit tests for the old implementation, just to make sure we don't break it in the future (coverage decreased, I guess because of this).

OK.

> Do you think it would make sense to include bitarray (mentioned in #51) as a dependency? Or binascii.hexlify/unhexlify? Or binhex.binhex/hexbin?

My opinion on this is that the fewer dependencies the better. Everything already works without it.

**Comment by djunzu** [comment]
ping

@JohannesBuchner, will you merge it? Any more change needed?

**Comment by JohannesBuchner** [comment]
Aha, sorry, I did not see that you added another commit. Thank you, I merged it. Could you both check if the newest version (4.0) works OK for you?

**Comment by djunzu** [comment]
Working!
🎉

**Pull request merged**